File: comixed-services/src/main/java/org/comixedproject/service/auditlog/RestAuditLogService.java
/*
 * ComiXed - A digital comic book library management application.
 * Copyright (C) 2020, The ComiXed Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses>
 */

package org.comixedproject.service.auditlog;

import java.util.Date;
import java.util.List;
import lombok.extern.log4j.Log4j2;
import org.comixedproject.model.auditlog.RestAuditLogEntry;
import org.comixedproject.repositories.auditlog.RestAuditLogRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * <code>RestAuditLogService</code> provides business logic services for instances of {@link
 * RestAuditLogEntry}.
 *
 * @author Darryl L. Pierce
 */
@Service
@Log4j2
public class RestAuditLogService {
  private static final Object SEMAPHORE = new Object();

  @Autowired private RestAuditLogRepository auditLogRepository;

  /**
   * Save a new entry to the log.
   *
   * @param entry the entry
   * @return the saved entry
   */
  @Transactional
  public RestAuditLogEntry save(final RestAuditLogEntry entry) {
    log.debug("Saving new REST audit log entry: started={}", entry.getStartTime());
    return this.auditLogRepository.save(entry);
  }

  /**
   * Returns all REST audit log entries after the specified threshold date.
   *
   * @param cutoff the cutoff timestamp
   * @return the entries
   */
  public List<RestAuditLogEntry> getEntriesAfterDate(final Long cutoff) {
    log.debug("Fetching up to 100 REST audit log entries");

    List<RestAuditLogEntry> result = null;
    boolean done = false;
    long started = System.currentTimeMillis();

    while (!done) {
      result =
          this.auditLogRepository.findByEndTimeAfterOrderByEndTime(
              new Date(cutoff), PageRequest.of(0, 100));
      if (result.isEmpty()) {
        log.debug("Waiting for REST audit log entries");
        synchronized (SEMAPHORE) {
          try {
            SEMAPHORE.wait(1000L);
          } catch (InterruptedException error) {
            log.debug("Interrupted while getting REST audit log entries", error);
            Thread.currentThread().interrupt();
          }
        }
      }

      done = !result.isEmpty() || (System.currentTimeMillis() - started > 60000);
    }

    return result;
  }
}

File: comixed-frontend/src/app/backend-status/services/rest-audit-log.service.ts
/*
 * ComiXed - A digital comic book library management application.
 * Copyright (C) 2019, The ComiXed Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses>
 */

import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { LoggerService } from '@angular-ru/logger';
import { GET_REST_AUDIT_LOG_ENTRIES_URL } from 'app/backend-status/backend-status.constants';
import { interpolate } from 'app/app.functions';

@Injectable({
  providedIn: 'root'
})
export class RestAuditLogService {
  constructor(private logger: LoggerService, private http: HttpClient) {}

  /**
   * Loads all entries after the specified cutoff date.
   *
   * @param cutoff the cutoff date
   */
  loadEntries(cutoff: number): Observable<any> {
    this.logger.debug('service: getting REST audit log entries:', cutoff);
    return this.http.get(
      interpolate(GET_REST_AUDIT_LOG_ENTRIES_URL, { cutoff: cutoff })
    );
  }
}
